src/main/java/io/cos/cas/osf/web/flow/login/OsfDefaultLoginPreparationAction.java

```diff
@@ -49,45 +50,63 @@ public OsfDefaultLoginPreparationAction(
     @Override
     protected Event doExecute(RequestContext context) {
 
-        OsfCasLoginContext loginContext;
-        final String serviceUrl = getEncodedServiceUrlFromRequestContext(context);
         final boolean institutionLogin = isInstitutionLogin(context);
         final String institutionId = getInstitutionIdFromRequestContext(context);
         final boolean orcidRedirect = isOrcidLoginAutoRedirect(context);
         final String orcidLoginUrl = getOrcidLoginUrlFromFlowScope(context);
 
+        final String encodedServiceUrl = getEncodedServiceUrlFromRequestContext(context);
+        final boolean defaultService = isFromFlowlessErrorPage(context);
+        final OsfUrlProperties osfUrl = Optional.of(context).map(
+                requestContext -> (OsfUrlProperties) requestContext.getFlowScope().get(OsfCasWebflowConstants.FLOW_PARAMETER_OSF_URL)
+        ).orElse(null);
+        if (osfUrl == null) {
+            LOGGER.error("The login web flow has not been initialized correctly.");
+            return error();
+        }
+        final String defaultServiceUrl = osfUrl.constructDefaultServiceUrl();
+
+        OsfCasLoginContext loginContext;
         loginContext = Optional.of(context).map(requestContext
                 -> (OsfCasLoginContext) requestContext.getFlowScope().get(PARAMETER_LOGIN_CONTEXT)).orElse(null);
         if (loginContext == null) {
             loginContext = new OsfCasLoginContext(
-                    serviceUrl,
+                    encodedServiceUrl,
                     institutionLogin,
                     institutionId,
                     orcidRedirect,
-                    orcidLoginUrl
+                    orcidLoginUrl,
+                    defaultService,
+                    defaultServiceUrl
             );
         } else {
-            loginContext.setServiceUrl(serviceUrl);
+            loginContext.setDefaultServiceUrl(encodedServiceUrl);
```

Comment:
Thanks for pointing this out @felliott . This should be `setEncodedServiceUrl()` instead.